### PR TITLE
Add Table Extension

### PR DESCRIPTION
**Pull request by iloveivyxuan**
This PR is the table extension.

Following is the summary of what I've done in this PR
- Align all APIs with desktop's table extension
- Have the same input type checkings and exceptions with desktop
- Add thorough tests, including all tests from desktop's [Table-Extension](https://github.com/NetLogo/Table-Extension/blob/hexy/tests.txt), except for TablesEquality tests

I didn't implement `groupAgents()` API. I got stuck on implementing it and would be appreciated if I can get an idea or information.

I've got some observations. 
I noticed that `groupAgents()` means to group agents by their attributes, like `color`, `xcor`, `size`...
After I `console.log(reporter)`, I found it's an anonymous function 
```
() { return SelfManager.self().getVariable("color"); }
```

I guess `SelfManager.self()` need to be replaced by `agent`. But I don't know how. I was confused about how to write js to implement something like `[color] of turtles`.

#### Other Notes
I didn't add `TableEquality` tests that are about to compare extensions objects. I'd like to open a new PR to implement this feature on table, matrix, array ...

### CODE REVIEW

File: engine/src/main/coffee/extensions/table.coffee [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+      return
+
+    # (Table, Any) => Any
+    get = (table, key) ->
+      checkInput({table: table})
+      if key not instanceof Array
+        return table.get(key)
+
+      origin_key = getOriginKey(table, key)
+      if origin_key?
```
Comment (LaCuneta) [inline engine/src/main/coffee/extensions/table.coffee:89]: We prefer camel case style for variable names, so this would be `originKey`.

**Comment by iloveivyxuan** [comment]
Thank you so much for the detailed information and the tip for faster debugging! That helps a lot. I find I can run my function without running `Galapagos`. This saves a lot of time for me!

I've fixed the code. Sorry for the late commitments.

File: engine/src/main/coffee/extensions/table.coffee [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+      group = new Map()
+
+      agentset.shufflerator().forEach( (agent) ->
+        key = workspace.world.selfManager.askAgent(reporter)(agent)
+        value = group.get(key) ? []
+        value.push(agent)
+        group.set(key, value)
+      )
+
+      group.forEach( (value, key, map) ->
```
Comment (LaCuneta) [inline engine/src/main/coffee/extensions/table.coffee:173]: Since `map` isn't used here you can just omit it, `(value, key) -> ...`

Comment (iloveivyxuan) [reply engine/src/main/coffee/extensions/table.coffee]: Okay, I've fixed it.

**Comment by LaCuneta** [comment]
Awesome, thanks again for all your work!

**Pull request merged**